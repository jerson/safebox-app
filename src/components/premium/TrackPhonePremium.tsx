import React, { useState, useEffect } from 'react';
import {
  StyleSheet,
  ViewStyle,
  View,
  StyleProp,
  TextStyle
} from 'react-native';
import Colors from '../../modules/constants/Colors';
import Text from '../ui/Text';
import Client from '../../services/Client';
import { EnableLocationRequest } from '../../proto/services_pb';
import Strings from '../../modules/format/Strings';
import Button from '../ui/Button';
import TextInput from '../ui/TextInput';
import useTextInput from '../hooks/useTextInput';
import ItemPremium from './ItemPremium';
import AlertMessage from '../ui/AlertMessage';
import Emitter from '../../modules/listener/Emitter';

const styles = StyleSheet.create({
  alertMessage: {
    marginTop: 10,
    marginLeft: 50,
    marginRight: 20
  } as ViewStyle,
  button: {
    height: 45,
    width: 45,
    paddingLeft: 15
  } as ViewStyle,
  buttonEnable: {
    height: 45,
    width: 45,
    paddingLeft: 15,
    marginTop: 5
  } as ViewStyle,
  label: { color: Colors.grey5 } as TextStyle,
  value: { color: Colors.grey6 } as TextStyle,
  purchasedContainer: {
    marginTop: 10,
    paddingLeft: 50,
    paddingRight: 20,
    flexDirection: 'row',
    alignItems: 'flex-start'
  } as ViewStyle,
  disableInfo: {
    flex: 1,
    paddingRight: 10,
    alignSelf: 'center'
  } as ViewStyle,
  disableContainer: {
    justifyContent: 'center',
    flexDirection: 'row'
  } as ViewStyle,
  enableContainer: {
    justifyContent: 'center',
    flexDirection: 'row'
  } as ViewStyle,
  textInputContainer: {
    flex: 1,
    marginBottom: 0,
    marginTop: 0,
    marginRight: 10
  } as ViewStyle
});

export interface TrackPhonePremiumProps {
  style?: StyleProp<ViewStyle>;
}

function TrackPhonePremium({ style }: TrackPhonePremiumProps) {
  const [isPurchased, setIsPurchased] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [email, setEmail] = useState('');
  const [emailInput, emailInputProps] = useTextInput('');
  const [error, setError] = useState('');

  const onPurchase = (isPurchased: boolean) => {
    setIsPurchased(isPurchased);
  };

  useEffect(() => {
    Emitter.emit('onTrackPhoneEnabled', !!email);
  }, [email]);

  const loadEmail = async () => {
    try {
      const response = await Client.getEmail();
      setEmail(response.getEmail());
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
  };

  useEffect(() => {
    if (!isPurchased) {
      setEmail('');
      return;
    }
    loadEmail();
  }, [isPurchased]);

  const disable = async () => {
    setIsLoading(true);
    try {
      await Client.disableLocation();
      setEmail('');
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
    setIsLoading(false);
  };

  const enable = async () => {
    setIsLoading(true);
    try {
      const request = new EnableLocationRequest();
      request.setEmail(emailInput);
      await Client.enableLocation(request);
      setEmail(emailInput);
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
    setIsLoading(false);
  };
  return (
    <ItemPremium
      style={style}
      onPurchase={onPurchase}
      productId={'trackphone'}
      name={'Track Phone'}
      icon={'map-pin'}
      description={`If you enable this option we will send you an email daily with the last location you used to connect to the application.`}
    >
      {!!error && (
        <AlertMessage
          onTimeout={() => {
            setError('');
          }}
          style={styles.alertMessage}
          message={error}
        />
      )}
      {isPurchased && (
        <View style={styles.purchasedContainer}>
          {!email && (
            <View style={styles.enableContainer}>
              <TextInput
                icon={'at-sign'}
                placeholder={'Insert you email'}
                defaultValue={''}
                keyboardType={'email-address'}
                autoCapitalize={'none'}
                autoCorrect={false}
                autoCompleteType={'email'}
                returnKeyType={'done'}
                onSubmitEditing={enable}
                containerStyle={styles.textInputContainer}
                {...emailInputProps}
              />
              <Button
                style={styles.buttonEnable}
                onPress={enable}
                isLoading={isLoading}
                typeColor={'primaryLight'}
                icon={'save'}
              />
            </View>
          )}
          {!!email && (
            <View style={styles.disableContainer}>
              <View style={styles.disableInfo}>
                <Text style={styles.label}>We will send your location to:</Text>
                <Text style={styles.value}>{email}</Text>
              </View>
              <Button
                style={styles.button}
                onPress={disable}
                isLoading={isLoading}
                typeColor={'accentDark'}
                icon={'trash-2'}
              />
            </View>
          )}
        </View>
      )}
    </ItemPremium>
  );
}

export default TrackPhonePremium;
