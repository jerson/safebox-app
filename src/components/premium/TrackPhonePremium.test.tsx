import React from 'react';
import * as renderer from 'react-test-renderer';
import TrackPhonePremium, {TrackPhonePremiumProps} from './TrackPhonePremium';

const props: TrackPhonePremiumProps = {};
test('renders without crashing', () => {
  const snapshot = renderer.create(<TrackPhonePremium {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<TrackPhonePremium {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
