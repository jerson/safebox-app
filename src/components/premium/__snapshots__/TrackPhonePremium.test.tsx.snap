// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
      },
      undefined,
    ]
  }
>
  <_class
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "marginBottom": 5,
        },
      ]
    }
  >
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#9066f5",
            "height": 1,
            "left": 0,
            "opacity": 0.5,
            "position": "absolute",
            "right": 0,
            "top": 12,
          },
          Object {
            "backgroundColor": "#c2d0e7",
          },
          false,
        ]
      }
    />
    <_class
      style={
        Array [
          Object {
            "backgroundColor": "#4531cd",
            "borderRadius": 10,
            "paddingBottom": 4,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 6,
          },
          Object {
            "backgroundColor": "#6781a6",
          },
          false,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#FFFFFF",
              "fontSize": 13,
              "lineHeight": 14,
              "minWidth": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Track Phone - Premium
      </Text>
    </_class>
  </_class>
  <_class
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 10,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#65748b",
              "fontSize": 28,
              "marginLeft": 10,
              "marginRight": 15,
            },
          ],
          Object {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ï†Ž
    </Text>
    <_class
      style={
        Object {
          "flex": 1,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Nunito",
              "fontSize": undefined,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {},
            Object {
              "color": "#65748b",
              "fontSize": 13,
              "marginBottom": 10,
            },
          ]
        }
      >
        If you enable this option we will send you an email daily with the last location you used to connect to the application.
      </Text>
      <_class
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#9066f5",
            "borderColor": "#9066f5",
            "borderRadius": 10,
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minHeight": 30,
            "opacity": 1,
            "padding": 12,
            "zIndex": 1,
          }
        }
        testID="button_touchable"
      >
        <_class
          style={
            Array [
              Object {
                "bottom": 0,
                "justifyContent": "center",
                "left": 10,
                "position": "absolute",
                "top": 0,
              },
              undefined,
              Object {
                "left": 0,
                "position": "relative",
              },
            ]
          }
        >
          <_class
            style={
              Array [
                false,
                false,
                false,
                Object {
                  "marginRight": 4,
                },
              ]
            }
          >
            <ActivityIndicator
              animating={true}
              color="#FFFFFF"
              hidesWhenStopped={true}
              size="small"
            />
          </_class>
        </_class>
      </_class>
    </_class>
  </_class>
</_class>
`;
